File: ocsgw/src/main/java/org/ostelco/ocsgw/datasource/local/LocalDataSource.java
package org.ostelco.ocsgw.datasource.local;

import org.jdiameter.api.IllegalDiameterStateException;
import org.jdiameter.api.InternalException;
import org.jdiameter.api.OverloadException;
import org.jdiameter.api.RouteException;
import org.jdiameter.api.cca.ServerCCASession;
import org.ostelco.diameter.CreditControlContext;
import org.ostelco.diameter.model.*;
import org.ostelco.ocs.api.CreditControlRequestType;
import org.ostelco.ocsgw.OcsServer;
import org.ostelco.ocsgw.datasource.DataSource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;


/**
 * Local DataSource that will accept all Credit Control Requests
 * Can be used as a bypass.
 *
 */
public class LocalDataSource implements DataSource {

    private static final Logger LOG = LoggerFactory.getLogger(LocalDataSource.class);

    @Override
    public void init() {
        // No init needed
    }

    @Override
    public void handleRequest(CreditControlContext context) {
        CreditControlAnswer answer = createCreditControlAnswer(context);
        LOG.info("Got Credit-Control-Request [{}] [{}]", context.getCreditControlRequest().getMsisdn(), context.getSessionId());
        try {
            final ServerCCASession session = OcsServer.INSTANCE.getStack().getSession(context.getSessionId(), ServerCCASession.class);
            session.sendCreditControlAnswer(context.createCCA(answer));
            LOG.info("Sent Credit-Control-Answer [{}] [{}]", context.getCreditControlRequest().getMsisdn(), context.getSessionId());
        } catch (InternalException | IllegalDiameterStateException | RouteException | OverloadException | NullPointerException e) {
            LOG.error("Failed to send Credit-Control-Answer. [{}] [{}]", context.getCreditControlRequest().getMsisdn(), context.getSessionId(), e);
        }
    }

    private CreditControlAnswer createCreditControlAnswer(CreditControlContext context) {

        final List<MultipleServiceCreditControl> origMultipleServiceCreditControls = context.getCreditControlRequest().getMultipleServiceCreditControls();
        final List<MultipleServiceCreditControl> newMultipleServiceCreditControls = new ArrayList<>();

        for (MultipleServiceCreditControl mscc : origMultipleServiceCreditControls) {

            FinalUnitIndication finalUnitIndication = null;

            if (context.getOriginalCreditControlRequest().getRequestTypeAVPValue() == CreditControlRequestType.TERMINATION_REQUEST.getNumber()) {
                finalUnitIndication = new FinalUnitIndication(
                        FinalUnitAction.TERMINATE,
                        new ArrayList<>(),
                        new ArrayList<>(),
                        new RedirectServer(
                                RedirectAddressType.IPV4_ADDRESS,
                                "")
                );
            }

            final List<ServiceUnit> newRequested = new ArrayList<>();
            for (ServiceUnit requested : mscc.getRequested()) {
                newRequested.add(new ServiceUnit(requested.getTotal(), 0, 0));
            }

            if (!newRequested.isEmpty()) {
                final ServiceUnit granted = newRequested.get(0);
                MultipleServiceCreditControl newMscc = new MultipleServiceCreditControl(
                        mscc.getRatingGroup(),
                        mscc.getServiceIdentifier(),
                        newRequested,
                        mscc.getUsed(),
                        granted,
                        mscc.getValidityTime(),
                        7200,
                        (long) (granted.getTotal() * 0.2), // 20%
                        finalUnitIndication,
                        ResultCode.DIAMETER_SUCCESS);

                newMultipleServiceCreditControls.add(newMscc);

            }
        }

        int validityTime = 0;
        if (newMultipleServiceCreditControls.isEmpty()) {
            validityTime = 86400;
        }

        return new CreditControlAnswer(ResultCode.DIAMETER_SUCCESS, newMultipleServiceCreditControls, validityTime);
    }

    public boolean isBlocked(final String msisdn) {
        return false;
    }
}

File: ocsgw/src/main/kotlin/org/ostelco/ocsgw/datasource/local/LocalDataSource.kt
package org.ostelco.ocsgw.datasource.local

import org.jdiameter.api.IllegalDiameterStateException
import org.jdiameter.api.InternalException
import org.jdiameter.api.OverloadException
import org.jdiameter.api.RouteException
import org.jdiameter.api.cca.ServerCCASession
import org.ostelco.diameter.CreditControlContext
import org.ostelco.diameter.model.*
import org.ostelco.ocs.api.CreditControlRequestType
import OcsServer.stack
import org.ostelco.diameter.getLogger
import org.ostelco.ocsgw.datasource.DataSource
import java.util.*

/**
 * Local DataSource that will accept all Credit Control Requests
 * Can be used as a bypass.
 */
class LocalDataSource : DataSource {

    private val logger by getLogger()

    override fun init() { // No init needed
    }

    override fun handleRequest(context: CreditControlContext) {
        val answer = createCreditControlAnswer(context)
        logger.info("Got Credit-Control-Request [{}] [{}]", context.creditControlRequest.msisdn, context.sessionId)
        try {
            val session = stack!!.getSession(context.sessionId, ServerCCASession::class.java)
            session.sendCreditControlAnswer(context.createCCA(answer))
            logger.info("Sent Credit-Control-Answer [{}] [{}]", context.creditControlRequest.msisdn, context.sessionId)
        } catch (e: InternalException) {
            logger.error("Failed to send Credit-Control-Answer. [{}] [{}]", context.creditControlRequest.msisdn, context.sessionId, e)
        } catch (e: IllegalDiameterStateException) {
            logger.error("Failed to send Credit-Control-Answer. [{}] [{}]", context.creditControlRequest.msisdn, context.sessionId, e)
        } catch (e: RouteException) {
            logger.error("Failed to send Credit-Control-Answer. [{}] [{}]", context.creditControlRequest.msisdn, context.sessionId, e)
        } catch (e: OverloadException) {
            logger.error("Failed to send Credit-Control-Answer. [{}] [{}]", context.creditControlRequest.msisdn, context.sessionId, e)
        } catch (e: NullPointerException) {
            logger.error("Failed to send Credit-Control-Answer. [{}] [{}]", context.creditControlRequest.msisdn, context.sessionId, e)
        }
    }

    override fun isBlocked(msisdn: String): Boolean {
        return false
    }

    private fun createCreditControlAnswer(context: CreditControlContext): CreditControlAnswer {
        val origMultipleServiceCreditControls = context.creditControlRequest.multipleServiceCreditControls
        val newMultipleServiceCreditControls: MutableList<MultipleServiceCreditControl> = ArrayList()
        for (mscc in origMultipleServiceCreditControls) {
            var finalUnitIndication: FinalUnitIndication? = null
            if (context.originalCreditControlRequest.requestTypeAVPValue == CreditControlRequestType.TERMINATION_REQUEST.number) {
                finalUnitIndication = FinalUnitIndication(
                        FinalUnitAction.TERMINATE,
                        ArrayList(),
                        ArrayList(),
                        RedirectServer(RedirectAddressType.IPV4_ADDRESS,"")
                )
            }
            val newRequested: MutableList<ServiceUnit> = ArrayList()
            for (requested in mscc.requested) {
                newRequested.add(ServiceUnit(requested.total, 0, 0))
            }
            if (!newRequested.isEmpty()) {
                val granted = newRequested[0]
                val newMscc = MultipleServiceCreditControl(
                        mscc.ratingGroup,
                        mscc.serviceIdentifier,
                        newRequested,
                        mscc.used,
                        granted,
                        mscc.validityTime,
                        7200,
                        (granted.total * 0.2).toLong(),  // 20%
                        finalUnitIndication,
                        ResultCode.DIAMETER_SUCCESS)
                newMultipleServiceCreditControls.add(newMscc)
            }
        }
        var validityTime = 0
        if (newMultipleServiceCreditControls.isEmpty()) {
            validityTime = 86400
        }
        return CreditControlAnswer(ResultCode.DIAMETER_SUCCESS, newMultipleServiceCreditControls, validityTime)
    }
}